gno.land/pkg/gnoweb/markdown/ext_forms.go

```diff
@@ -0,0 +1,350 @@
+		}
+
+		if strings.TrimSpace(result.Name) == "" {
+			// If "name" is missing, it's an error, but keep the type
+			result.Error = ErrFormInputMissingName
+			return result
+		}
+
+		// Validate input type
+		if !validateInputType(result.InputType) {
+			result.InputType = defaultInputType
+			result.Error = ErrFormInvalidInputType
+			return result
+		}
+
+		return result
+	}
+
+	// Extract path attribute for form tag
+	if bytes.HasPrefix(trimmed, []byte("<gno-form")) {
+		result.Tag = FormTagOpen
+		result.RenderPath = extractAttr(trimmed, "path")
+		return result
+	}
+
+	result.Error = ErrFormUnexpectedOrInvalidTag
+	return result
+}
+
+// extractAttr extracts the value of the first found attribute from a line
+func extractAttr(line []byte, attr string) string {
+	pattern := attr + "=\""
+	if i := bytes.Index(line, []byte(pattern)); i != -1 {
+		start := i + len(pattern)
+		if end := bytes.IndexByte(line[start:], '"'); end != -1 {
+			return string(line[start : start+end])
+		}
+	}
+	return ""
+}
+
+// --- The Block Parser ---
+
+// formParser starts a block as soon as we encounter "<gno-form>"
+// and closes it as soon as we encounter "</gno-form>".
+// In between, only <gno-input /> lines are processed.
+type formParser struct{}
+
+var _ parser.BlockParser = (*formParser)(nil)
+
+// NewFormParser creates a new instance of formParser
+func NewFormParser() *formParser {
+	return &formParser{}
+}
+
+// Trigger detects the start of the block
+func (p *formParser) Trigger() []byte {
+	return []byte{'<'}
+}
+
+// Open starts a block only when the line is exactly "<gno-form>"
+func (p *formParser) Open(parent ast.Node, reader text.Reader, pc parser.Context) (ast.Node, parser.State) {
+	line, seg := reader.PeekLine()
+	info := parseFormTag(line)
+	if info.Error != nil || info.Tag != FormTagOpen {
+		return nil, parser.NoChildren
+	}
+	node := NewForm(FormTagOpen)
+	node.RenderPath = info.RenderPath
+	// Consume the line "<gno-form>"
+	reader.Advance(seg.Len())
+	return node, parser.HasChildren
+}
+
+// Continue processes lines until "</gno-form>" is encountered.
+// When a line contains <gno-input />, it adds a child node.
+func (p *formParser) Continue(node ast.Node, reader text.Reader, pc parser.Context) parser.State {
+	line, seg := reader.PeekLine()
+	if seg.Len() == 0 {
+		return parser.Close
+	}
+	trimmed := bytes.TrimSpace(line)
+	// If the line is exactly "</gno-form>", consume it and close the block.
+	if bytes.Equal(trimmed, []byte("</gno-form>")) {
+		reader.Advance(seg.Len())
+		return parser.Close
+	}
+	// If the line starts with "<gno-input", we add an input node.
+	info := parseFormTag(line)
+	if info.Tag == FormTagInput {
+		input := NewForm(FormTagInput)
+
+		// Get realm name and placeholder
+		input.InputName = info.Name
+		input.InputType = info.InputType
+		if info.Placeholder == "" {
+			input.InputPlaceholder = defaultPlaceholder
+		} else {
+			input.InputPlaceholder = info.Placeholder
+		}
```

Comment:
nit: can this be simplified?

```suggestion
		input.InputPlaceholder = info.Placeholder
		if info.Placeholder == "" {
			input.InputPlaceholder = defaultPlaceholder
		} 
```